### PR TITLE
Adding run combinators for cli and daemon

### PR DESCRIPTION
**Pull request by nhibberd**
! @charleso @jystic @thumphries @tmcgilchrist 

Ideal use case

```
# cli
main :: IO ()
main = do
  cli "my-cli" buildInfoVersion dependencyInfo myThingParser $ \c -> 
    case c of
      DoThingA -> ...
      DoThingB -> ...

# daemon
main :: IO ()
main = do
  daemon "my-daemon" buildInfoVersion dependencyInfo $ do
    loop my-loop
    
```

### CODE REVIEW

**Comment by nhibberd** [comment]
Have separated the name / version to clean up the interface

**Comment by markhibberd** [comment]
:+1: looks useful/neat, would be good to have at least a smoke test as well.

File: x-optparse/src/X/Options/Applicative.hs [hunk 1]
```diff
@@ -86,6 +90,24 @@ dispatch p = getArgs >>= \x -> case x of
             >>= handleParseResult . removeError
   _  -> execParser (info (p <**> helper) idm)
 
+cli :: Show a => [Char] -> [Char] -> [[Char]] -> Parser a -> (a -> IO b) -> IO b
```
Comment (thumphries) [inline x-optparse/src/X/Options/Applicative.hs]: newtypes? or at least a usage example for this function from strings and strings and strings to things

**Comment by thumphries** [comment]
Thanks for the docs 📖

**Pull request merged**